### PR TITLE
Add shim for Python's gRPC core on Windows

### PR DESCRIPTION
**Pull request by soltanmm**
This PR isn't so much a pull request as a plea/request.

cc @nicolasnoble @ctiller

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -189,7 +196,16 @@ def cython_extensions(package_names, module_names, include_dirs, libraries,
 TEST_LOADER = 'tests:Loader'
 TEST_RUNNER = 'tests:Runner'
 
-PACKAGE_DATA = {}
+PACKAGE_DATA = {
+    'grpc._adapter': [
+        'credentials/roots.pem'
+    ],
+    'grpc._cython': [
+        '_windows/grpc.def',
```
Comment (nicolasnoble) [inline setup.py]: That file's not needed. Trash it.

File: setup.py [hunk 2]
```diff
@@ -75,14 +75,23 @@
...
+if not "win32" in sys.platform:
+    CORE_C_FILES += tuple(grpc_core_dependencies.CORE_SOURCE_FILES)
+
 EXTENSION_INCLUDE_DIRECTORIES = (
     (PYTHON_STEM,) + CORE_INCLUDE + BORINGSSL_INCLUDE + ZLIB_INCLUDE)
 
 EXTENSION_LIBRARIES = ('m',)
-if not "darwin" in sys.platform:
-  EXTENSION_LIBRARIES += ('rt',)
+if not "darwin" in sys.platform and not "win32" in sys.platform:
```
Comment (nicolasnoble) [inline setup.py]: `if "linux" in sys.platform` instead.

File: src/python/grpcio/grpc/_cython/_windows/grpc.def [hunk 3]
```diff
@@ -0,0 +1,261 @@
+EXPORTS
```
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/_windows/grpc.def]: Trash that file.

File: src/python/grpcio/grpc/_cython/loader.c [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+int pygrpc_load_core(const char *path) {
+  if (!SetDllDirectoryA(path)) {
+    return 0;
+  }
+#if GPR_ARCH_64
+  TCHAR fname[] = _T("grpc_c.64.python");
+#else
+  TCHAR fname[] = _T("grpc_c.32.python");
+#endif
+  HMODULE module = GetModuleHandle(_T("grpc_c.so"));
```
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/loader.c]: That `grpc_c.so` has to be the name of the dll that is your extension.

File: src/python/grpcio/grpc/_cython/loader.c [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#if GPR_WIN32
```
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/loader.c]: You removed the inclusion to port_platform.h. So GPR_WIN32 isn't defined, that condition is false, and you're skipping the whole windows loading code to fallback to the default "yay everything is fine" at the bottom. That's why you're segfaulting. You haven't loaded anything, all of your symbols are NULL.

**Comment by nicolasnoble** [comment]
Also, obviously, do not let the compiled DLLs in there. We'll generate them for you beforehand as an artifact from another machine. But now we know you need these in `src/python/grpcio/grpc/_cython/_windows/grpc_c.{32,64}.python` :-)

File: setup.py [hunk 6]
```diff
@@ -189,7 +196,16 @@ def cython_extensions(package_names, module_names, include_dirs, libraries,
...
 
-PACKAGE_DATA = {}
+PACKAGE_DATA = {
+    'grpc._adapter': [
+        'credentials/roots.pem'
+    ],
+    'grpc._cython': [
+        '_windows/grpc.def',
+        '_windows/grpc_c.32.python',
+        '_windows/grpc_c.64.python',
```
Comment (nicolasnoble) [inline setup.py]: Does that mean you're going to ship the Windows dlls with all the packages at all times ? Linux and MacOS included ? And source ?

File: src/python/grpcio/grpc/_cython/cygrpc.pyx [hunk 7]
```diff
@@ -45,6 +48,10 @@ include "grpc/_cython/_cygrpc/server.pyx.pxi"
 cdef class _ModuleState:
 
   def __cinit__(self):
+    filename = pkg_resources.resource_filename(__name__, '_windows/grpc.def')
```
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/cygrpc.pyx]: The Def file is irrelevant. However if you do know the absolute location of the dll, that simplifies things.

File: src/python/grpcio/grpc/_cython/loader.c [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+#include <tchar.h>
+
+int pygrpc_load_core(const char *path) {
+  if (!SetDllDirectoryA(path)) {
```
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/loader.c]: Why?

File: src/python/grpcio/grpc/_cython/loader.c [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  TCHAR fname[] = _T("grpc_c.32.python");
+#endif
+  HMODULE module = GetModuleHandle(_T("grpc_c.so"));
+  TCHAR path[2048 + 32] = _T("");
+  LPTSTR seek_back = NULL;
+  GetModuleFileName(module, path, 2048);
+
+  seek_back = _tcsrchr(path, _T('\\'));
+
+  while (seek_back) {
```
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/loader.c]: If you have the path already, just load the thing directly, avoid these shenanigans. That is needed for Ruby because we can't know where we are installed.

Use LoadLibraryA if you pass a `char *` path.

File: src/python/grpcio/grpc/_cython/loader.c (hunk 8 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/loader.c]: Because I have no idea how Windows works, don't feel like taking the time to learn whatever subtleties may be involved, and want the DLLs to be found.

File: src/python/grpcio/grpc/_cython/loader.c (hunk 4 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/loader.c]: Platform specific funtime...

File: setup.py (hunk 2 above)
Comment (soltanmm) [inline setup.py]: Do other `*nix`s not need it?

File: setup.py (hunk 6 above)
Comment (soltanmm) [inline setup.py]: Yes, and I'm going to avoid thinking about it too much until this frenetic rush to getting Windows support subsides.

File: setup.py (hunk 2 above)
Comment (nicolasnoble) [inline setup.py]: librt is a Linux aberration only, yes.

File: setup.py (hunk 6 above)
Comment (nicolasnoble) [inline setup.py]: That's fine. Trying to understand where we're going and what will be left afterward.

File: src/python/grpcio/grpc/_cython/loader.c (hunk 8 above)
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/loader.c]: So... you don't know how Windows works, and you're trying to alleviate your fear of having the platform break by adding more unknown function calls to it ? I fail to follow the logic here :-)

File: src/python/grpcio/grpc/_cython/loader.c (hunk 4 above)
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/loader.c]: So, what is the name of the dll created by the python compiler ? That's what you have to put here instead.

**Comment by nicolasnoble** [comment]
See soltanmm/grpc#2. It's almost what you want to do.

File: src/python/grpcio/grpc/_cython/loader.c (hunk 8 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/loader.c]: I don't like the approach being taken here. The better approach would have been to figure out what the difference between MSYS2's emulation of Linux and Linux actual was to then have the entire thing build within setuptools. Instead, we scrapped trying to dig too deep on that one. This is merely a continuation of that reasoning.

File: src/python/grpcio/grpc/_cython/loader.h [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ */
+
+#ifndef PYGRPC_LOADER_H_
+#define PYGRPC_LOADER_H_
+
+#include "imports.generated.h"
+
+/* Additional inclusions not covered by "imports.generated.h" */
+#include <grpc/byte_buffer_reader.h>
```
Comment (nicolasnoble) [inline src/python/grpcio/grpc/_cython/loader.h]: Wait, why is that necessary ?

Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/loader.h]: (responded to offline)
Summary: wanted to keep all the header inclusions used by the extension library in one place; `loader.h` seemed like a decent place to keep it that way. Will add explanatory comment later.
